File: BasicVoxelEngine/InputManager.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

namespace BasicVoxelEngine
{
    [StructLayout(LayoutKind.Sequential)]
    public struct KeyState
    {
        public bool Up { get; set; }
        public bool Down { get; set; }
        public bool Held { get; set; }
    }
    ///<summary>
    /// probably unnecessary
    ///</summary>
    public enum Key
    {
        UNKNOWN = -1,
        SPACE = 32,
        APOSTROPHE = 39,
        COMMA = 44,
        MINUS = 45,
        PERIOD = 46,
        SLASH = 47,
        N0 = 48,
        N1 = 49,
        N2 = 50,
        N3 = 51,
        N4 = 52,
        N5 = 53,
        N6 = 54,
        N7 = 55,
        N8 = 56,
        N9 = 57,
        SEMICOLON = 59,
        EQUAL = 61,
        A = 65,
        B = 66,
        C = 67,
        D = 68,
        E = 69,
        F = 70,
        G = 71,
        H = 72,
        I = 73,
        J = 74,
        K = 75,
        L = 76,
        M = 77,
        N = 78,
        O = 79,
        P = 80,
        Q = 81,
        R = 82,
        S = 83,
        T = 84,
        U = 85,
        V = 86,
        W = 87,
        X = 88,
        Y = 89,
        Z = 90,
        LEFT_BRACKET = 91,
        BACKSLASH = 92,
        RIGHT_BRACKET = 93,
        GRAVE_ACCENT = 96,
        WORLD_1 = 161,
        WORLD_2 = 162,
        ESCAPE = 256,
        ENTER = 257,
        TAB = 258,
        BACKSPACE = 259,
        INSERT = 260,
        DELETE = 261,
        RIGHT = 262,
        LEFT = 263,
        DOWN = 264,
        UP = 265,
        PAGE_UP = 266,
        PAGE_DOWN = 267,
        HOME = 268,
        END = 269,
        CAPS_LOCK = 280,
        SCROLL_LOCK = 281,
        NUM_LOCK = 282,
        PRINT_SCREEN = 283,
        PAUSE = 284,
        F1 = 290,
        F2 = 291,
        F3 = 292,
        F4 = 293,
        F5 = 294,
        F6 = 295,
        F7 = 296,
        F8 = 297,
        F9 = 298,
        F10 = 299,
        F11 = 300,
        F12 = 301,
        // lets just skip all other F keys
        KP_0 = 320,
        KP_1 = 321,
        KP_2 = 322,
        KP_3 = 323,
        KP_4 = 324,
        KP_5 = 325,
        KP_6 = 326,
        KP_7 = 327,
        KP_8 = 328,
        KP_9 = 329,
        KP_DECIMAL = 330,
        KP_DIVIDE = 331,
        KP_MULTIPLY = 332,
        KP_SUBTRACT = 333,
        KP_ADD = 334,
        KP_ENTER = 335,
        KP_EQUAL = 336,
        LEFT_SHIFT = 340,
        LEFT_CONTROL = 341,
        LEFT_ALT = 342,
        LEFT_SUPER = 343,
        RIGHT_SHIFT = 344,
        RIGHT_CONTROL = 345,
        RIGHT_ALT = 346,
        RIGHT_SUPER = 347,
        MENU = 348
    }
    public sealed class InputManager
    {
        public const int LeftMouseButton = 0;
        public const int RightMouseButton = 1;
        public const int MiddleMouseButton = 2;
        internal InputManager(IntPtr address)
        {
            mAddress = address;
        }
        ~InputManager()
        {
            DestroyRef_Native(mAddress);
        }
        public bool MouseEnabled
        {
            get => GetMouseEnabled_Native(mAddress);
            set => SetMouseEnabled_Native(mAddress, value);
        }
        public Vector2 Mouse => GetMouse_Native(mAddress);
        public KeyState GetKey(Key key) => GetKey_Native(mAddress, key);
        public KeyState GetMouseButton(int button) => GetMouseButton_Native(mAddress, button);
        private readonly IntPtr mAddress;
        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void DestroyRef_Native(IntPtr address);
        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern bool GetMouseEnabled_Native(IntPtr address);
        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void SetMouseEnabled_Native(IntPtr address, bool value);
        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern Vector2 GetMouse_Native(IntPtr address);
        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern KeyState GetKey_Native(IntPtr address, Key key);
        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern KeyState GetMouseButton_Native(IntPtr address, int button);
    }
}

File: src/lib/input_manager.cpp
#include "bve_pch.h"
#include "input_manager.h"
namespace bve {
#define key(KEY) GLFW_KEY_##KEY
    static const std::vector<int32_t> keys = {
        key(ESCAPE), key(F1), key(F2), key(F3), key(F4), key(F5), key(F6), key(F7), key(F8), key(F9), key(F10), key(F11), key(F12),
        key(GRAVE_ACCENT), key(1), key(2), key(3), key(4), key(5), key(6), key(7), key(8), key(9), key(0), key(MINUS), key(EQUAL), key(BACKSPACE),
        key(TAB), key(Q), key(W), key(E), key(R), key(T), key(Y), key(U), key(I), key(O), key(P), key(LEFT_BRACKET), key(RIGHT_BRACKET), key(BACKSLASH),
        key(CAPS_LOCK), key(A), key(S), key(D), key(F), key(G), key(H), key(J), key(K), key(L), key(SEMICOLON), key(APOSTROPHE), key(ENTER),
        key(LEFT_SHIFT), key(Z), key(X), key(C), key(V), key(B), key(N), key(M), key(COMMA), key(PERIOD), key(SLASH), key(RIGHT_SHIFT),
        key(LEFT_CONTROL), key(LEFT_SUPER), key(LEFT_ALT), key(SPACE), key(RIGHT_ALT), key(RIGHT_SUPER), key(RIGHT_CONTROL)
    };
#undef key
    static std::map<GLFWwindow*, input_manager*> window_map;
    input_manager::input_manager(ref<window> window_) {
        this->m_mouse = this->m_last_mouse = this->m_current_offset = glm::vec2(0.f);
        this->m_window = window_;
        this->m_mouse_enabled = false;
        window_map.insert({ this->m_window->m_window, this });
        glfwSetCursorPosCallback(this->m_window->m_window, mouse_callback);
        for (int32_t key : keys) {
            key_state state;
            state.reset();
            this->m_states.insert({ key, state });
        }
        for (size_t i = 0; i < 8; i++) {
            key_state state;
            state.reset();
            this->m_mouse_buttons.push_back(state);
        }
    }
    input_manager::~input_manager() {
        window_map.erase(this->m_window->m_window);
    }
    input_manager::key_state input_manager::get_key(int32_t glfw_key) {
        return this->m_states[glfw_key];
    }
    input_manager::key_state input_manager::get_mouse_button(size_t button) {
        return this->m_mouse_buttons[button];
    }
    glm::vec2 input_manager::get_mouse() {
        return this->m_current_offset;
    }
    static void finalize_state(input_manager::key_state& current, const input_manager::key_state& old) {
        current.down = current.held && !old.held;
        current.up = !current.held && old.held;
    }
    void input_manager::update() {
        glfwSetInputMode(this->m_window->m_window, GLFW_CURSOR, this->m_mouse_enabled ? GLFW_CURSOR_NORMAL : GLFW_CURSOR_DISABLED);
        this->m_current_offset = (this->m_mouse - this->m_last_mouse) * glm::vec2(1.f, -1.f);
        this->m_last_mouse = this->m_mouse;
        for (int32_t key : keys) {
            key_state old_state = this->m_states[key];
            key_state& state = this->m_states[key];
            state.reset();
            state.held = glfwGetKey(this->m_window->m_window, key) == GLFW_PRESS;
            finalize_state(state, old_state);
        }
        for (size_t i = 0; i < this->m_mouse_buttons.size(); i++) {
            key_state old_state = this->m_mouse_buttons[i];
            key_state& state = this->m_mouse_buttons[i];
            state.reset();
            state.held = glfwGetMouseButton(this->m_window->m_window, (int32_t)i) == GLFW_PRESS;
            finalize_state(state, old_state);
        }
    }
    bool& input_manager::mouse_enabled() {
        return this->m_mouse_enabled;
    }
    void input_manager::mouse_callback(GLFWwindow* glfw_window, double x, double y) {
        input_manager* im = window_map[glfw_window];
        im->m_mouse = glm::vec2((float)x, (float)y);
    }
}